defi/src/adaptors/handlers/storeAdaptorData/index.ts

```diff
@@ -206,19 +212,49 @@ export const handler2 = async (event: IStoreAdaptorDataHandlerEvent) => {
       } else
         throw new Error("Invalid adapter")
 
+
+      if (runType === 'store-all') {
+        let runAtCurrTime = adaptersToRun.some(([_version, adapter]) => Object.values(adapter).some(a => a.runAtCurrTime))
+
+        const date = new Date()
+        const hours = date.getUTCHours()
+        const isExpensiveAdapter = adaptor.isExpensiveAdapter
+        // if it is an expensive adapter run every 4 hours or after 20:00 UTC
+        const runNow = !isExpensiveAdapter || (hours % 4 === 0 || hours > 20)
+        const haveTodayData = todayIdSet.has(id)
+        const haveYesterdayData = yesterdayIdSet.has(id)
+
+
+        if (runAtCurrTime || !isAdapterVersionV1) {
+          recordTimestamp = timestampAtStartofHour
+          endTimestamp = timestampAtStartofHour
+
+          if (haveTodayData && !runNow) {
+            console.info(`Skipping ${adapterType} - ${index + 1}/${protocols.length} - ${protocol.module} - skipping, already have today data for adapter running at current time`)
+            return
+          }
+        } else { // it is a version 1 adapter - we pull yesterday's data
+          if (haveYesterdayData) {
+            console.info(`Skipping ${adapterType} - ${index + 1}/${protocols.length} - ${protocol.module} - skipping, already have yesterday data`)
+            return;
+          }
+
+          endTimestamp = getTimestampAtStartOfDayUTC(Math.floor(Date.now() / 1000) - 1)
+          recordTimestamp = getTimestampAtStartOfDayUTC(endTimestamp)
+        }
+      }
+
       const promises: any = []
       const rawRecords: RawRecordMap = {}
       const adaptorRecords: {
         [key: string]: AdaptorRecord
       } = {}
       for (const [version, adapter] of adaptersToRun) { // the version is the key for the record (like uni v2) not the version of the adapter
-        const runAtCurrTime = Object.values(adapter).some(a => a.runAtCurrTime)
-        // if (runAtCurrTime && Math.abs(LAMBDA_TIMESTAMP - toTimestamp) > 60 * 60 * 3)
-        //   throw new Error('This Adapter can be run only around current time') // allow run current time if within 3 hours
         const chainBlocks = {} // WARNING: reset chain blocks for each adapter, sharing this between v1 & v2 adapters that have different end timestamps have nasty side effects
         const runAdapterRes = await runAdapter(adapter, endTimestamp, chainBlocks, module, version, { adapterVersion })
-        // const runAdapterRes = await runAdapterInSubprocess({ adapter, endTimestamp, chainBlocks, module, version, adapterVersion })
 
+        const recordWithTimestamp = runAdapterRes.find((r: any) => r.timestamp)
+        if (recordWithTimestamp) recordTimestamp = recordWithTimestamp.timestamp
```

Comment:
if a single adapter returns a timestamp we change the recorded timestamp for all adapters?

Reply:
recordTimestamp is localized to each adapter run

Reply:
true, my bad here